flow/JsonTraceLogFormatter.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+ * JsonTraceLogFormatter.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2018 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "flow/flow.h"
+#include "flow/JsonTraceLogFormatter.h"
+
+#include <sstream>
+
+void JsonTraceLogFormatter::addref() {
+	ReferenceCounted<JsonTraceLogFormatter>::addref();
+}
+
+void JsonTraceLogFormatter::delref() {
+	ReferenceCounted<JsonTraceLogFormatter>::delref();
+}
+
+const char* JsonTraceLogFormatter::getExtension() {
+	return "json";
+}
+
+const char* JsonTraceLogFormatter::getHeader() {
+	return "";
+}
+
+const char* JsonTraceLogFormatter::getFooter() {
+	return "";
+}
+
+namespace {
+
+void escapeString(std::stringstream& ss, const std::string source) {
+	for (auto c : source) {
+		if (c == '"') {
+			ss << "\\\"";
+		} else if (c == '\\') {
+			ss << "\\\\";
+		} else {
+			ss << c;
+		}
+	}
+}
+
+} // namespace
+
+std::string JsonTraceLogFormatter::formatEvent(const TraceEventFields& fields) {
+	std::stringstream ss;
+	ss << "{  ";
+	for (auto iter = fields.begin(); iter != fields.end(); ++iter) {
+		if (iter != fields.begin()) {
+			ss << ", ";
+		}
+		ss << "\"";
+		escapeString(ss, iter->first);
+		ss << "\": \"";
+		escapeString(ss, iter->second);
+		ss << "\"";
+	}
+	ss << " }\r\n";
```

Comment:
Is this generating a file with multiple root elements? My gut says that our output should be a valid JSON document, which we could do by adding a `{` header, a `}` footer, and commas separating the events.

Reply:
It's not a valid json document exactly. We're generating line-delimited json [1] so that we can parse it line-wise which is more memory efficient. It works well with tools like jq [2]

[1]: https://en.wikipedia.org/wiki/JSON_streaming#Line-delimited_JSON
[2]: https://stedolan.github.io/jq/

Reply:
Hmm, that does seems defensible, especially when also combined with the fact that a log file that hasn't been closed yet (or ever) will be invalid with only the header. As long as it's well supported by existing tools, it seems reasonable.

Reply:
The drawback of this approach is that it is inconsistent with the xml formatter. But honestly I never liked the header and footer of FDBs XML format, I would have preferred a document per line.

One thing we could do to make this more consistent is have a knob whether a whole file should be a document or whether it should be by line. In XMLs case we would remove header/footer if the knob is set to doc-per-line, for JSON we would start the trace-file with `[` and end it with `]` if a document per file is requested.

What I don't like about this is that FDB can't guarantee that a trace file will be a valid XML/JSON document anyways, so I prefer the per-line approach.

Reply:
Ah yes, it would be array brackets rather than the object brackets I suggested. I don't know if I feel strongly that the XML and JSON formats have to follow the same approach in this regard, but I can appreciate the appeal of having each event be a distinct document on its own line. I can see if anybody else around here has an opinion.